### PR TITLE
feat: add leader info to UI

### PR DESCRIPTION
**Pull request by Cifko**
Description
---
Add leader info to UI (leader, leader round). Also to each qc add leader round and proposed_by to see where the QC come from.

How Has This Been Tested?
---
Visually.

### CODE REVIEW

File: dan_layer/storage_sqlite/src/sqlite_shard_store_factory.rs [hunk 1]
```diff
@@ -970,6 +1002,54 @@ impl ShardStoreWriteTransaction<PublicKey, TariDanPayload> for SqliteShardStoreW
...
+        };
+
+        match diesel::insert_into(current_states::table)
+            .values(&new_row)
+            .execute(self.connection())
+        {
+            Ok(_) => {},
+            Err(err) => match err {
+                Error::DatabaseError(DatabaseErrorKind::UniqueViolation, _) => {
+                    match diesel::update(current_states::table)
```
Comment (jorgeantonio21) [inline dan_layer/storage_sqlite/src/sqlite_shard_store_factory.rs]: I think you can do something more idiomatic, as follows:

```rs
diesel::insert_into(current_states::table)
    .values(&new_row)
    .on_conflict(id)
    .do_update()
    .set(current_states::leader_round.eq(i64::from(leader_round)))
    .filter(current_states::payload_id.eq(payload_id))
    .execute(&conn);
 ```

**Review by jorgeantonio21 — DISMISSED** [review]

File: dan_layer/storage_sqlite/src/sqlite_shard_store_factory.rs (hunk 1 above)
Comment (Cifko) [reply dan_layer/storage_sqlite/src/sqlite_shard_store_factory.rs]: If we ever update diesel to 2.x then I can change this. The on_conflict is supported only for PG, not sqlite.

File: dan_layer/core/src/storage/shard_store.rs [hunk 2]
```diff
@@ -163,6 +165,13 @@ pub trait ShardStoreWriteTransaction<TAddr: NodeAddressable, TPayload: Payload>
     fn rollback(self) -> Result<(), StorageError>;
     fn insert_high_qc(&mut self, from: TAddr, shard: ShardId, qc: QuorumCertificate) -> Result<(), StorageError>;
     fn save_payload(&mut self, payload: TPayload) -> Result<(), StorageError>;
+    fn save_current_state(
```
Comment (sdbondi) [inline dan_layer/core/src/storage/shard_store.rs]: nit: perhaps `save_current_leader_state`

File: dan_layer/core/src/storage/shard_store.rs [hunk 3]
```diff
@@ -115,6 +116,7 @@ pub trait ShardStoreReadTransaction<TAddr: NodeAddressable, TPayload: Payload> {
     fn get_high_qcs(&self, payload_id: PayloadId) -> Result<Vec<QuorumCertificate>, StorageError>;
     /// Returns the current leaf node for the shard
     fn get_leaf_node(&self, payload_id: &PayloadId, shard: &ShardId) -> Result<LeafNode, StorageError>;
+    fn get_current_states(&self, payload: &PayloadId) -> Result<Vec<SQLCurrentState>, StorageError>;
```
Comment (sdbondi) [inline dan_layer/core/src/storage/shard_store.rs]: The store interface trait shouldn't know about "SQL" - the intermediate SQLite structs should be kept in the storage_sqlite crate and converted to models in core as needed. e.g. shard_id has to be a Vec<u8> for SQL but should be `ShardId` in the public interface in `core` and converted in storage_sqlite

Should name this something more specific than "state" `get_current_leader_states` or `get_current_hotstuff_round_states`

**Review by sdbondi — COMMENTED** [review]
Looks good, would like a more descriptive name for the JRPC endpoint, table name, functions and structs and avoiding the use of SQL in the core storage layer since we will be implementing the trait for LMDB in future

File: dan_layer/core/src/storage/shard_store.rs (hunk 3 above)
Comment (Cifko) [reply dan_layer/core/src/storage/shard_store.rs]: It's not SQL, I just didn't know what to name it :-) so I took similar approach as get_transaction (and now I realized that I put that in). Argh, I'm very bad at naming...
Renamed

Comment (sdbondi) [reply dan_layer/core/src/storage/shard_store.rs]: haha no problem :)

**Review by stringhandler — COMMENTED** [review]
tested but was not showing recent transactions. I fixed that and then saw this on the page: 

```
Unhandled Thrown Error!
Cannot read properties of undefined (reading 'substring')
TypeError: Cannot read properties of undefined (reading 'substring')
    at qe (http://localhost:5000/static/js/main.9be2dd6e.js:2:363877)
    at http://localhost:5000/static/js/main.9be2dd6e.js:2:453614
    at Array.map (<anonymous>)
    at Bs (http://localhost:5000/static/js/main.9be2dd6e.js:2:453330)
    at Co (http://localhost:5000/static/js/main.9be2dd6e.js:2:212611)
    at Sl (http://localhost:5000/static/js/main.9be2dd6e.js:2:269853)
    at xc (http://localhost:5000/static/js/main.9be2dd6e.js:2:258789)
    at vc (http://localhost:5000/static/js/main.9be2dd6e.js:2:258717)
    at yc (http://localhost:5000/static/js/main.9be2dd6e.js:2:258580)
    at oc (http://localhost:5000/static/js/main.9be2dd6e.js:2:255360)
💿 Hey developer 👋
```

**Comment by Cifko** [comment]
> tested but was not showing recent transactions. I fixed that and then saw this on the page:
> 
> ```
> Unhandled Thrown Error!
> Cannot read properties of undefined (reading 'substring')
> TypeError: Cannot read properties of undefined (reading 'substring')
>     at qe (http://localhost:5000/static/js/main.9be2dd6e.js:2:363877)
>     at http://localhost:5000/static/js/main.9be2dd6e.js:2:453614
>     at Array.map (<anonymous>)
>     at Bs (http://localhost:5000/static/js/main.9be2dd6e.js:2:453330)
>     at Co (http://localhost:5000/static/js/main.9be2dd6e.js:2:212611)
>     at Sl (http://localhost:5000/static/js/main.9be2dd6e.js:2:269853)
>     at xc (http://localhost:5000/static/js/main.9be2dd6e.js:2:258789)
>     at vc (http://localhost:5000/static/js/main.9be2dd6e.js:2:258717)
>     at yc (http://localhost:5000/static/js/main.9be2dd6e.js:2:258580)
>     at oc (http://localhost:5000/static/js/main.9be2dd6e.js:2:255360)
> 💿 Hey developer 👋
> ```

It's super weird, it works when I run the web ui separately, but not from the app...

**Comment by Cifko** [comment]
Now it works, at least on my machine on a fresh setup. We should think about adding UI tests. Because the UI was broken before this PR.

**Review by sdbondi — DISMISSED** [review]

**Review by sdbondi — DISMISSED** [review]
Thanks

Looks like the test is broken on development :/ will try and see whats up

**Review by stringhandler — APPROVED** [review]

**Pull request merged**